_pages/exercises/AutonomousCars/end_to_end_visual_control.md

```diff
@@ -0,0 +1,414 @@
+To address this, you should first <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">categorize</code> the samples into their respective classes and then apply data balancing techniques, such as:
+
+  <ul>
+    <li><strong>Undersampling:</strong> Reducing the number of majority-class samples.</li>
+    <li><strong>Oversampling:</strong> Increasing the number of minority-class samples, sometimes by duplicating or augmenting them.</li>
+  </ul>
+</p>
+
+<p style="text-align:justify">
+In the preprocessing stage, images are first loaded from the dataset folders and paired with their corresponding labels from <strong>CSV files</strong> that contain control values such as <strong>linear</strong> and <strong>angular velocity</strong>. To ensure consistency, all images are <strong>resized</strong> to a fixed dimension and <strong>normalized</strong> by scaling pixel values to a 
+standard range, which improves training stability. The <strong>labels</strong> are extracted directly from the CSV files and linked to the appropriate images. To enhance <strong>robustness</strong> and <strong>generalization</strong>, <strong>data augmentation techniques</strong> 
+such as <strong>flipping</strong>, <strong>rotation</strong>, and <strong>brightness adjustment</strong> are applied, 
+creating diverse variations of the training data.
+</p>
+
+#### ii) Dataset Splitting
+
+<p style="text-align:justify">
+The <strong>Simple Circuit dataset</strong> is divided into three sets: <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">training</code> and <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">testing</code>. The training set contains the majority of the data and is used to teach the model to recognize patterns and learn the task. The testing set is the data used during training to fine-tune hyper-parameters, prevent over-fitting, and monitor the model’s performance on unseen data.
+</p>
+
+<p style="text-align:justify;">
+The <strong>Combine Circuit Dataset</strong> consists of <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">training</code> data and <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">adjustment data</code> (used for sharp turns). If needed, these can be combined to create a larger dataset. You can then split the combined dataset into <strong>training</strong> and <strong>testing</strong> sets using an <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">80%-20% ratio</code> with the <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">train_test_split</code> function <code class="language-plaintext highlighter-rouge" style="color:#222831;background:#bdbdbd">(from sklearn.model_selection)</code>.
```

Comment:
Same here, it is more correct to say that "You can then split the combined dataset into training and **validation** sets". Again, we don't have a true test set, our test is done directly in simulation.